Project list is fetched every time from every host

Need to pull the projects list out to its own HashMap (if possible)
that is then passed around and checked.

As pushed in #3, the HashMap for project_list has been extracted, but causes the following errors on `cargo check`.

```
error[E0596]: cannot borrow `project_list` as mutable, as it is not declared as mutable
  --> src/main.rs:60:42
   |
46 | fn get_necessary_data_from_hosts(project_list : &mut HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
   |                                  ------------ help: consider changing this to be mutable: `mut project_list`
...
60 |         update_projects_list(&mut hosts, &mut project_list).await;
   |                                          ^^^^^^^^^^^^^^^^^ cannot borrow as mutable

error[E0596]: cannot borrow `project_list` as mutable, as it is not declared as mutable
   --> src/main.rs:275:47
    |
271 | fn get_data_for_model(store : &gtk::ListStore, project_list : &mut HashMap<Option<String>, models::ProjectInfo>) {
    |                                                ------------ help: consider changing this to be mutable: `mut project_list`
...
275 |     let hosts = get_necessary_data_from_hosts(&mut project_list);
    |                                               ^^^^^^^^^^^^^^^^^ cannot borrow as mutable

error: aborting due to 2 previous errors
```

As of 00f17d1, we now have the following on `cargo check`

```
warning: variable does not need to be mutable
  --> src/main.rs:66:55
   |
66 | async fn update_projects_list(hosts : &mut Vec<Host>, mut project_list : &HashMap<Option<String>, models::ProjectInfo>)
   |                                                       ----^^^^^^^^^^^^
   |                                                       |
   |                                                       help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

error[E0596]: cannot borrow `*project_list` as mutable, as it is behind a `&` reference
   --> src/main.rs:103:17
    |
103 |                 project_list.insert(project.url, my_proj);
    |                 ^^^^^^^^^^^^ `project_list` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```

--- a/src/main.rs
+++ b/src/main.rs
@@ -43,7 +43,7 @@ impl Host {
     }
 }
 
-fn get_necessary_data_from_hosts(mut project_list : &HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
+fn get_necessary_data_from_hosts(mut project_list : &mut HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
 {
     let mut hosts = Vec::new();
 
@@ -57,13 +57,13 @@ fn get_necessary_data_from_hosts(mut project_list : &HashMap<Option<String>, mod
     });
 
     tokio::runtime::Runtime::new().unwrap().block_on(async {
-        update_projects_list(&mut hosts, &mut project_list).await;
+        update_projects_list(&hosts, &mut project_list).await;
     });
 
     hosts
 }
 
-async fn update_projects_list(hosts : &mut Vec<Host>, mut project_list : &HashMap<Option<String>, models::ProjectInfo>)
+async fn update_projects_list(hosts : &Vec<Host>, mut project_list : &mut HashMap<Option<String>, models::ProjectInfo>)
 {
     for (_idx, host) in hosts.into_iter().enumerate() {
         // Here we want to check to see if we need to fetch
@@ -268,7 +268,7 @@ fn build_ui(application: &gtk::Application) {
     });
 }
 
-fn get_data_for_model(store : &gtk::ListStore, mut project_list : &HashMap<Option<String>, models::ProjectInfo>) {
+fn get_data_for_model(store : &gtk::ListStore, mut project_list : &mut HashMap<Option<String>, models::ProjectInfo>) {
     store.clear();
 
     let col_indices: [u32; 14] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
@@ -279,7 +279,7 @@ fn get_data_for_model(store : &gtk::ListStore, mut project_list : &HashMap<Optio
             for (_, result) in results.into_iter().enumerate() {
                 let values: [&dyn ToValue; 14] = [
                     &host.addr,
-                    &host.projects[&result.project_url].name,
+                    &project_list.get(&result.project_url).unwrap().name,
                     &result.name,
                     &format!("{0:.2} %", result.progress()),
                     &result.elapsed_as_string(),

--- a/src/main.rs
+++ b/src/main.rs
@@ -43,7 +43,7 @@ impl Host {
     }
 }
 
-fn get_necessary_data_from_hosts(project_list : &mut HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
+fn get_necessary_data_from_hosts(mut project_list : &HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
 {
     let mut hosts = Vec::new();
 
@@ -63,7 +63,7 @@ fn get_necessary_data_from_hosts(project_list : &mut HashMap<Option<String>, mod
     hosts
 }
 
-async fn update_projects_list(hosts : &mut Vec<Host>, project_list : &mut HashMap<Option<String>, models::ProjectInfo>)
+async fn update_projects_list(hosts : &mut Vec<Host>, mut project_list : &HashMap<Option<String>, models::ProjectInfo>)
 {
     for (_idx, host) in hosts.into_iter().enumerate() {
         // Here we want to check to see if we need to fetch
@@ -73,17 +73,17 @@ async fn update_projects_list(hosts : &mut Vec<Host>, project_list : &mut HashMa
         // list - if so, we need to fetch the proejcts list
         // from the host
 
-        let mut has_all_projects : bool = false;
+        let mut has_missing_projects : bool = false;
 
         if let Some(results) = host.results.clone() {
             for (_, result) in results.into_iter().enumerate() {
                 if ! project_list.contains_key(&result.project_url) {
-                    has_all_projects = true;
+                    has_missing_projects = true;
                 }
             }
         }
 
-        if ! has_all_projects {
+        if ! has_missing_projects {
             return;
         }
 
@@ -268,7 +268,7 @@ fn build_ui(application: &gtk::Application) {
     });
 }
 
-fn get_data_for_model(store : &gtk::ListStore, project_list : &mut HashMap<Option<String>, models::ProjectInfo>) {
+fn get_data_for_model(store : &gtk::ListStore, mut project_list : &HashMap<Option<String>, models::ProjectInfo>) {
     store.clear();
 
     let col_indices: [u32; 14] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

--- a/src/main.rs
+++ b/src/main.rs
@@ -43,7 +43,7 @@ impl Host {
     }
 }
 
-fn get_necessary_data_from_hosts() -> Vec<Host>
+fn get_necessary_data_from_hosts(project_list : &mut HashMap<Option<String>, models::ProjectInfo>) -> Vec<Host>
 {
     let mut hosts = Vec::new();
 
@@ -57,13 +57,13 @@ fn get_necessary_data_from_hosts() -> Vec<Host>
     });
 
     tokio::runtime::Runtime::new().unwrap().block_on(async {
-        update_projects_list(&mut hosts).await;
+        update_projects_list(&mut hosts, &mut project_list).await;
     });
 
     hosts
 }
 
-async fn update_projects_list(hosts : &mut Vec<Host>)
+async fn update_projects_list(hosts : &mut Vec<Host>, project_list : &mut HashMap<Option<String>, models::ProjectInfo>)
 {
     for (_idx, host) in hosts.into_iter().enumerate() {
         // Here we want to check to see if we need to fetch
@@ -77,7 +77,7 @@ async fn update_projects_list(hosts : &mut Vec<Host>)
 
         if let Some(results) = host.results.clone() {
             for (_, result) in results.into_iter().enumerate() {
-                if ! host.projects.contains_key(&result.project_url) {
+                if ! project_list.contains_key(&result.project_url) {
                     has_all_projects = true;
                 }
             }
@@ -100,7 +100,7 @@ async fn update_projects_list(hosts : &mut Vec<Host>)
         if let Some(projects) = client_projects.clone() {
             for (_, project) in projects.into_iter().enumerate() {
                 let my_proj = project.clone();
-                host.projects.insert(project.url, my_proj);
+                project_list.insert(project.url, my_proj);
             }
         }
     }
@@ -200,6 +200,7 @@ fn create_model() -> gtk::ListStore {
 
 fn build_ui(application: &gtk::Application) {
     let model = Rc::new(RefCell::new(create_model()));
+    let mut project_list : HashMap<Option<String>, models::ProjectInfo> = HashMap::new();
 
     let window = gtk::ApplicationWindow::new(application);
     window.set_title("BOINCView");
@@ -235,13 +236,13 @@ fn build_ui(application: &gtk::Application) {
 
     add_columns(&treeview);
 
-    get_data_for_model(&model.borrow());
+    // get_data_for_model(&model.borrow());
 
     // Every 5 seconds, we'll update the data
     Some(gtk::timeout_add(
         5000,
         move || {
-            get_data_for_model(&model.borrow());
+            get_data_for_model(&model.borrow(), &mut project_list);
 
             glib::Continue(true)
         }
@@ -267,11 +268,11 @@ fn build_ui(application: &gtk::Application) {
     });
 }
 
-fn get_data_for_model(store : &gtk::ListStore) {
+fn get_data_for_model(store : &gtk::ListStore, project_list : &mut HashMap<Option<String>, models::ProjectInfo>) {
     store.clear();
 
     let col_indices: [u32; 14] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
-    let hosts = get_necessary_data_from_hosts();
+    let hosts = get_necessary_data_from_hosts(&mut project_list);
 
     for host in hosts {
         if let Some(results) = host.results {

